clang/lib/Sema/SemaCoroutine.cpp

```diff
@@ -1965,9 +1965,15 @@ bool Sema::buildCoroutineParameterMoves(SourceLocation Loc) {
     if (PD->getType()->isDependentType())
       continue;
 
+    // Preserve the referenced state for unused parameter diagnostics.
+    bool DeclReferenced = PD->isReferenced();
+
     ExprResult PDRefExpr =
         BuildDeclRefExpr(PD, PD->getType().getNonReferenceType(),
                          ExprValueKind::VK_LValue, Loc); // FIXME: scope?
+
+    PD->setReferenced(DeclReferenced);
```

Comment:
What happens in `BuildDeclRefExpr` such that `PD` loses track of whether it is referenced? If something is changing `PD` in a bad way, perhaps at that point might be better to set the appropriated method instead.

Reply:
As you mentioned offline, it's because creating the expression will mark it referenced, gotcha.

Reply:
`BuildDeclRefExpr` builds a `DeclRef` `Expr` and marks the referenced `Decl` as `Referenced`.